### DISCUSSION TITLE
Why do we need `TrainState`?

### DISCUSSION THREAD

**Discussion by andsteing** [google/flax#1573] [Q&A] [answered]
Question that came up in discussion with @relational

It's not obvious how to use `TrainState.apply_gradients`:

https://github.com/google/flax/blob/d7172e9d3ea75a68a218ca68d07ee384d656664f/flax/training/train_state.py#L58-L72

Why not simply call `model.apply()` and what does the `update_step()` in the pydoc refer to?

**Comment by andsteing** [comment, marked as answer]
`update_step()` was referring to the update during the training step (renamed in #1279 to make this a bit clearer).

The main advantage of `TrainState` is that it handles Optax + param updates, which is the same pattern in 95% of cases. In the spirit of progressive disclosure of complexity we thought adding a small helper for specifically this pattern would be useful.

Note that especially for `train_step()` (aka `update_step()`) we usually already have a lot of arguments (inputs, targets, ...) so it is nice to bundle the `apply_fn()` together with the other state and transformations (like optimizer). See also #1444 for some more discussion on this.
                      
If you're interested in more details of the decision how to structure `TrainState` please refer to https://github.com/google/flax/blob/master/docs/flip/1009-optimizer-api.md

**Comment by andsteing** [comment]
What happens to these function attributes of the `TrainState` dataclass when the train state is checkpointed?

**Reply by andsteing** [reply]
Because these attributes are decorated with `struct.field(pytree_node=False)` they're neither transformed nor written to disc.

https://github.com/google/flax/blob/d7172e9d3ea75a68a218ca68d07ee384d656664f/flax/struct.py#L63-L65
